src/coreclr/vm/unsafeaccessors.cpp

```diff
@@ -75,16 +93,285 @@ namespace
 
         UnsafeAccessorKind Kind;
         MethodDesc* Declaration;
-        MetaSig DeclarationSig;
+        SigPointer DeclarationSig;  // This is the official declaration signature. It may be modified
+                                    // to include the UnsafeAccessorTypeAttribute types.
+        MetaSig DeclarationMetaSig;
+        NewArrayHolder<ParamDetails> TranslatedParams; // Redefined types from UnsafeAccessorTypeAttribute usage.
+                                                       // Return type is at 0 index. Function arguments are 1 to N.
         SigPointer TargetTypeSig;
         TypeHandle TargetType;
         bool IsTargetStatic;
         MethodDesc* TargetMethod;
         FieldDesc* TargetField;
     };
 
+    void AppendTypeToSignature(
+        SigBuilder& sig,
+        TypeHandle th)
+    {
+        STANDARD_VM_CONTRACT;
+        _ASSERTE(!th.IsNull());
+
+        //
+        // Building the signature follows details defined in ECMA-335 - II.23.2.12
+        //
+
+        CorElementType elemType = th.GetSignatureCorElementType();
+        if (CorIsPrimitiveType(elemType))
+        {
+            sig.AppendElementType(elemType);
+            return;
+        }
+
+        if (th.IsGenericVariable())
+        {
+            TypeVarTypeDesc* typeVar = th.AsGenericVariable();
+            sig.AppendElementType(typeVar->GetInternalCorElementType());
+            sig.AppendData(typeVar->GetIndex());
+            return;
+        }
+
+        if (th.HasTypeParam())
+        {
+            // Append the element type.
+            sig.AppendElementType(elemType);
+            TypeHandle typeParam = th.GetTypeParam();
+            AppendTypeToSignature(sig, typeParam);
+
+            // Append ArrayShape for MD arrays
+            // See II.23.2.13 ArrayShape
+            if (elemType == ELEMENT_TYPE_ARRAY)
+            {
+                DWORD rank = th.GetRank();
+                sig.AppendData(rank);
+                sig.AppendData(0); // Append the number of sizes.
+
+                // Roslyn always emits the lower bounds of 0 for each dimension.
+                // In order to match the signature, we also need to append the number
+                // of lower bounds for each dimension.
+                // We can emit 0 for each lower bound, since UnsafeAccessors is only
+                // supported in C# and C# doesn't support lower bounds.
+                sig.AppendData(rank);
+                for (DWORD i = 0; i < rank; i++)
+                    sig.AppendData(0);
+            }
+            return;
+        }
+
+        // Extract the instantiation from parameterized type.
+        MethodTable* pMT = th.GetMethodTableOfRootTypeParam();
```

Comment:
```suggestion
        MethodTable* pMT = th.GetMethodTable();
```
Param types were handled above